src/visitors/cvode_visitor.cpp

```diff
@@ -0,0 +1,154 @@
+        logger->debug("CvodeVisitor :: replacing {} with {} on LHS of {}",
+                      name->get_node_name(),
+                      varname,
+                      to_nmodl(node));
+        node.set_lhs(std::make_shared<ast::Name>(new ast::String(varname)));
+        if (program_symtab->lookup(varname) == nullptr) {
+            auto symbol = std::make_shared<symtab::Symbol>(varname, ModToken());
+            symbol->set_original_name(name->get_node_name());
+            program_symtab->insert(symbol);
+        }
+        if (block_index == BlockIndex::JACOBIAN) {
+            auto rhs = node.get_rhs();
+            // map of all indexed symbols (need special treatment in SymPy)
+            auto name_map = get_name_map(*rhs, name->get_node_name());
+            auto diff2c = pywrap::EmbeddedPythonLoader::get_instance().api().diff2c;
+            auto [jacobian,
+                  exception_message] = diff2c(to_nmodl(*rhs), name->get_node_name(), name_map);
+            if (!exception_message.empty()) {
+                logger->warn("CvodeVisitor :: python exception: {}", exception_message);
+            }
+            // NOTE: LHS can be anything here, the equality is to keep `create_statement` from
+            // complaining, we discard the LHS later
+            auto statement = fmt::format("{} = {} / (1 - dt * ({}))", varname, varname, jacobian);
+            logger->debug("CvodeVisitor :: replacing statement {} with {}",
+                          to_nmodl(node),
+                          statement);
+            auto expr_statement = std::dynamic_pointer_cast<ast::ExpressionStatement>(
+                create_statement(statement));
+            const auto bin_expr = std::dynamic_pointer_cast<const ast::BinaryExpression>(
+                expr_statement->get_expression());
+            node.set_rhs(std::shared_ptr<ast::Expression>(bin_expr->get_rhs()->clone()));
+        }
+    }
+}
+
+void CvodeVisitor::visit_program(ast::Program& node) {
+    program_symtab = node.get_symbol_table();
+    node.visit_children(*this);
+    if (derivative_block) {
+        auto der_node = new ast::CvodeBlock(derivative_block->get_name(),
+                                            derivative_block->get_statement_block(),
+                                            std::shared_ptr<ast::StatementBlock>(
+                                                derivative_block->get_statement_block()->clone()));
+        node.emplace_back_node(der_node);
+    }
+
+    // re-visit the AST since we now inserted the CVODE block
+    node.visit_children(*this);
+    if (!conserve_equations.empty()) {
+        auto blocks = collect_nodes(node, {ast::AstNodeType::CVODE_BLOCK});
+        auto block = std::dynamic_pointer_cast<ast::CvodeBlock>(blocks[0]);
+        block->get_function_block()->erase_statement(conserve_equations);
+        block->get_diagonal_jacobian_block()->erase_statement(conserve_equations);
+    }
+}
```

Comment:
This visitor does:
1. Collect CONSERVE statements (for later deletion).
2. Recurses once to create an invalid CVODE_BLOCK.
3. Keeps track of which "branch" of the invalid `CVODE_BLOCK` we're in via a counter `block_index`.
4. Keeps track of two additional layers booleans: `in_cvode_block`  and `in_differential_equation`.

When doing a:
``` 
node.visit_children(*this);
// changes to AST.
node.visit_children(*this);
```
it's not obvious which functionality is desired during the first pass; and what's desired during the second pass. It could be doing everything again, or one of the boolean flags turn off some functionality of the first/second pass.

I can't quickly tell if it's supposed to handle multiple DERIVATIVE blocks. ("No." is a good answer, see #1501). Note, it doesn't matter that it can't handle both, the issue is that it's hard to tell if it tries to.

I think there's room to make the intent/algorithm more visible. To my understanding the following is happening:
1. Find the one unique DERIVATIVE block.
2. Create a CVODE_BLOCK.
3. Insert the CVODE_BLOCK into the program.

The key is that you create a valid `CVODE_BLOCK` from the DERIVATIVE block. In pseudo code, I think you can do:
```
auto derivative_block = collect_nodes(node, "DERIVATIVE");
auto cvode_block = make_cvode_block(derivative_block);
node.emplace_back_node(derivative_block);
```
This change means, we can see the high-level idea in `CvodeVisitor`. The next problem is how to structure `make_cvode_block`.

I would propose something like:
```
CvodeBlock make_cvode_block(Derivative& derivative) {

    auto Dx_block = derivative.clone();
    Dx_block.accpet(CvodeDBlockVisitor());
    Dx_block.accept(RemoveConserveVisitor());

    auto J_block = derivative.clone();
    J_block.accept(CvodeJBlockVisitor());
    J_block.accept(RemoveConserveVisitor());

    return CVodeBlock(block_name, Dx_block, J_block);
}
```

Now it's clear that `make_cvode_block` simply creates the CVODE_BLOCK by transforming two copies of the derivative block. The transformation for each block can be written down in a visitor each, those are the poorly named `CvodeDxBlockVisitor` and `CvodeJBlockVisitor` (I suspect there's better names for both).

To get rid of CONSERVE blocks, you can (implement and) run a `RemoveConserveVisitor` on whatever node needs to be stripped.